### PR TITLE
cephfs: Add ceph-dokan, providing Windows support

### PR DESCRIPTION
**Pull request by petrutlucian94**
cephfs: Add ceph-dokan, providing Windows support

In order to expose ceph filesystems to Windows hosts, we propose
including ceph-dokan[1][2] in the Ceph tree, while updating it to
work with the latest CephFS and Dokany APIs.

Dokany is a well maintained project (fork of the original Dokan
project), allowing filesystems to be implemented in userspace,
even providing a Fuse compatibility layer.

One reason for not using the FUSE compatibility layer is that it's
only covering the high level API while Ceph is using the low level
FUSE API, which among other things is inode centric.

Changes made by this patch compared to the upstream ceph-dokan:
* support latest stable Dokany API. The upstream version relies on
  the legacy unmaintained Dokan API
* return proper error codes, converting standard errno.h values to
  NTSTATUS
* minor changes to support latest cephfs API
* drop duplicated ceph code, no longer needed if we're to include it
  in tree. This makes it much easier to maintain.
* drop redundant permission checks, leaving it up to libcephfs
* use ceph argparse helpers
* use ceph logging and daemon initialization
* fixed unicode handling
* switched to ceph coding style
* made ceph.conf param optional, using the default path if available
* enabled setting file timestamps
* append support
* configurable timeouts set once per mount
* ensure that the error code is always logged
* various cleanups (removed unused entry points, checks that have
  been moved to dokany, simplified conditional statements,
  unnecessary conversions in the hot path, etc).

[1] https://github.com/ketor/ceph-dokan
[2] https://github.com/ceph/ceph-dokan

Signed-off-by: Lucian Petrut <lpetrut@cloudbasesolutions.com>

Depends-On: https://github.com/ceph/ceph/pull/34858

Tracker issue: https://tracker.ceph.com/issues/49623

### CODE REVIEW

**Comment by petrutlucian94** [comment]
jenkins test make check

**Comment by petrutlucian94** [comment]
jenkins test make check

File: src/dokan/ceph_dokan.c
Comment (gregsfortytwo) [inline src/dokan/ceph_dokan.c:41]: These #define's need to get the boolean value enclosed in parentheses -- I know they're used safely here, but it'd be easy for somebody to stick them inside a compound statement and have very bad things happen.

File: src/dokan/posix_acl.h
Comment (gregsfortytwo) [inline src/dokan/posix_acl.h:1]: I am very confused by what's happening with the "posix acl" stuff here. It looks like it's independently re-implementing ACL semantics? Isn't there a library we can use so it's known-good?

**Review by gregsfortytwo — COMMENTED** [review]
Okay, so in looking at the structure of this I've got a few questions in individual comments.

But my biggest concern about a PR like this is systematic -- is there a testing plan to keep it live and functioning code? There's no tests included here; AFAIK the code isn't compiled by our CI system because we don't do Windows; and I'm not certain what a good integration test plan would look like. (It also doesn't comply with our code style rules, but given that it's inherited code written against a Windows API we'll just swallow that one. 😃) Has this been discussed elsewhere against the other Ceph systems? Is there a plan in progress?

**Review by petrutlucian94 — COMMENTED** [review]
@gregsfortytwo Thanks for feedback!

About the tests: for now, we'd like to run a Windows build before most if not all Ceph PRs that get merged. I've submitted a PR against ``ceph-build``, updating the ``ceph-dev-new`` job: https://github.com/ceph/ceph-build/pull/1722

Apart from that, we did port *some* of the Ceph tests, mostly covering librados and ceph-common: https://github.com/ceph/ceph/pull/34859. Hopefully, we might be able to port at least some of the already existing cephfs tests. Those tests (https://github.com/dimov-cz/winfstest) were also very useful while working on ceph-dokan.

We're still discussing the possibility of integrating Windows nodes with the Ceph CI so that we could actually run some Windows tests. Until that happens, we're going to run the tests on our own, probably as part of the pipeline that's building Windows Ceph MSIs: https://github.com/cloudbase/ceph-windows-installer. FWIW, I think BSD is in a similar situation.

> It also doesn't comply with our code style rules, but given that it's inherited code written against a Windows API we'll just swallow that one.

I'll be honest, RBD was our main focus. For cephfs, we saw that the ceph GH org already had a fork of ``ceph-dokan``, which was unmaintained but seemed promising and a good start, at least for V1. What we did here was to bring it up to date, fixing some issues along the way.

Here's our TODO list (some of which you've already mentioned):

* update the CLI, using the ceph argparsing helpers or boost. I'm thinking about printing a warning such as "This is a preview version of ceph-dokan, the CLI might change."
* symlinks
* use ceph coding style
* configurable volume name
* consider using a centralized service like the one we have for RBD, responsible of
managing the daemons. For now, NSSM might be an option.
* switch to standard ceph logging
* file lock support
* Windows ACL mapping (at least partial)
* cephfs testing, porting some of the existing tests and maybe add new ones as well

We're probably going to have a V2 including most of the above, depending on how much traction this will get. Even so, IMO it's good idea to include this code in-tree, making it easier to build and maintain and also more visible.

File: src/dokan/posix_acl.h
Comment (mattbenjamin) [reply src/dokan/posix_acl.h:1]: Not sure if it's relevant, but all Andreas' work seems to be lgpl 2.1: http://savannah.nongnu.org/projects/acl

Comment (petrutlucian94) [reply src/dokan/posix_acl.h:1]: Thanks for the suggestion, nice to see that's actually maintained. At a quick glance, I think it would require a few changes in order to build using Mingw but it's definitely worth considering, at least for the next iteration.

**Comment by petrutlucian94** [comment]
jenkins test api

**Comment by jan--f** [comment]
Looks like this is ready for another round of reviews. ping @batrick

**Comment by petrutlucian94** [comment]
I'll push a significant update to this PR this week, fixing a few issues and conforming to the rest of the ceph code in terms of coding style, arg parsing and logging.

File: src/dokan/README.md
Comment (petrutlucian94) [reply src/dokan/README.md]: I've updated the code to use the ``client`` ceph logging subsystem.

File: src/dokan/posix_acl.h
Comment (petrutlucian94) [reply src/dokan/posix_acl.h]: Even better, looks like we can drop all the permission checks performed at ceph-dokan level and leave that up to libcephfs.

The only edge case that we have is the delete operation, in which the ``Delete`` Dokan callback is supposed to tell if the delete operation is allowed, having the actual delete operation performed in the ``Cleanup`` callback.

**Comment by petrutlucian94** [comment]
Good news, I've pushed a major cleanup, reducing about half of the LOC. Here's what changed:

* dropped ceph-dokan redundant permission checks, relying on libcephfs instead
* disabled fake inodes. sizeof(ino_t) is 2 on Windows, so it was crashing after the first 65K inodes. Windows supports 64b identifiers through different structures, so fake inodes aren't needed
* switched to ceph coding style
* switched to ceph logging
* use ceph argparse helpers and clean up the syntax
* support the "--user" argument by using ``global_init``
* cleaned up log messages

**Comment by petrutlucian94** [comment]
jenkins test make check

**Comment by batrick** [comment]
@petrutlucian94 wasn't able to find you on IRC. Just want to let you know I won't be able to make the CDM tonight due to another commitment I forgot about. You can skip the CDM today just let Josh Durgin know you don't plan to attend (since it's early for you).

I think we will move forward towards merging this for Pacific. I will review in the next ~2 days.

File: src/common/win32/errno.cc
Comment (batrick) [inline src/common/win32/errno.cc:416]: As of fd04c30296899c8b9a0962edd14d26a11bf7abd8 (which will be backported to Pacific), there are now CephFS errnos which should be converted from instead.

File: src/dokan/README.md
Comment (batrick) [inline src/dokan/README.md:1]: This documentation should be in `doc/cephfs/` as an `.rst` with appropriate formatting. I would expect a basic usage section similar to `ceph-fuse`/`kernel`. I would also expect a small blurb about the windows driver here: https://docs.ceph.com/en/latest/cephfs/mount-prerequisites/

Comment (batrick) [reply src/dokan/README.md:1]: Following this, I'd encourage a blog post on ceph.io. Would you be interested?

**Review by batrick — CHANGES_REQUESTED** [review]
Otherwise looks good to merge.

File: src/common/win32/errno.cc
Comment (petrutlucian94) [reply src/common/win32/errno.cc:416]: Thanks for bringing it up. I'll add another function called ``cephfs_errno_to_ntsatus`` and use that instead.

File: src/dokan/README.md
Comment (petrutlucian94) [reply src/dokan/README.md:1]: > This documentation should be in `doc/cephfs/` as an `.rst` with appropriate formatting. I would expect a basic usage section similar to `ceph-fuse`/`kernel`. I would also expect a small blurb about the windows driver here: https://docs.ceph.com/en/latest/cephfs/mount-prerequisites/

Makes sense. TBH, I was thinking about having some developer notes in ``src/dokan/README.md`` and the install and usage guide in ``README.windows.rst`` but it's quite redundant ATM. I'll start preparing a `doc/cephfs/` doc.

Comment (petrutlucian94) [reply src/dokan/README.md:1]: > Following this, I'd encourage a blog post on ceph.io. Would you be interested?

That would be great, thanks! We actually got WP access yesterday so we'll prepare a blog post.

**Comment by petrutlucian94** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

**Pull request merged**
